src/apphosting/backend.ts

```diff
@@ -0,0 +1,482 @@
+import { promptOnce } from "../prompt";
+import { DEFAULT_LOCATION } from "./constants";
+import { ensure } from "../ensureApiEnabled";
+import * as deploymentTool from "../deploymentTool";
+import { DeepOmit } from "../metaprogramming";
+import { webApps } from "./app";
+import { GitRepositoryLink } from "../gcp/devConnect";
+import * as ora from "ora";
+import fetch from "node-fetch";
+import { orchestrateRollout } from "./rollout";
+
+const DEFAULT_COMPUTE_SERVICE_ACCOUNT_NAME = "firebase-app-hosting-compute";
+
+const apphostingPollerOptions: Omit<poller.OperationPollerOptions, "operationResourceName"> = {
+  apiOrigin: apphostingOrigin(),
+  apiVersion: API_VERSION,
+  masterTimeout: 25 * 60 * 1_000,
+  maxBackoff: 10_000,
+};
+
+async function tlsReady(url: string): Promise<boolean> {
+  // Note, we do not use the helper libraries because they impose additional logic on content type and parsing.
+  try {
+    await fetch(url);
+    return true;
+  } catch (err) {
+    // At the time of this writing, the error code is ERR_SSL_SSLV3_ALERT_HANDSHAKE_FAILURE.
+    // I've chosen to use a regexp in an attempt to be forwards compatible with new versions of
+    // SSL.
+    const maybeNodeError = err as { cause: { code: string }; code: string };
+    if (
+      /HANDSHAKE_FAILURE/.test(maybeNodeError?.cause?.code) ||
+      "EPROTO" === maybeNodeError?.code
+    ) {
+      return false;
+    }
+    return true;
+  }
+}
+
+async function awaitTlsReady(url: string): Promise<void> {
+  let ready;
+  do {
+    ready = await tlsReady(url);
+    if (!ready) {
+      await sleep(1000 /* ms */);
+    }
+  } while (!ready);
+}
+
+/**
+ * Set up a new App Hosting backend.
+ */
+export async function doSetup(
+  projectId: string,
+  webAppName: string | null,
+  location: string | null,
+  serviceAccount: string | null,
+): Promise<void> {
+  await Promise.all([
+    ensure(projectId, developerConnectOrigin(), "apphosting", true),
+    ensure(projectId, cloudbuildOrigin(), "apphosting", true),
+    ensure(projectId, secretManagerOrigin(), "apphosting", true),
+    ensure(projectId, cloudRunApiOrigin(), "apphosting", true),
+    ensure(projectId, artifactRegistryDomain(), "apphosting", true),
+    ensure(projectId, iamOrigin(), "apphosting", true),
+  ]);
+
+  // Hack: Because IAM can take ~45 seconds to propagate, we provision the service account as soon as
+  // possible to reduce the likelihood that the subsequent Cloud Build fails. See b/336862200.
+  await ensureAppHostingComputeServiceAccount(projectId, serviceAccount);
+
+  const allowedLocations = (await apphosting.listLocations(projectId)).map((loc) => loc.locationId);
+  if (location) {
+    if (!allowedLocations.includes(location)) {
+      throw new FirebaseError(
+        `Invalid location ${location}. Valid choices are ${allowedLocations.join(", ")}`,
+      );
+    }
+  }
+
+  location =
+    location || (await promptLocation(projectId, "Select a location to host your backend:\n"));
```

Comment:
Do these commands support noninteractive use cases? If so, we should double check this code - it seems like it will prompt even in noninteractive mode if there is no location.